index.js

```diff
@@ -1,33 +1,73 @@
 "use strict";
 
-function getNextAsyncIteratorValue(asyncIterator, index) {
-    return asyncIterator.next().then((iterator) => {
-        return { index, iterator };
-    });
+function getNextAsyncIteratorFactory(options) {
+    return async(asyncIterator, index) => {
+        try {
+            const iterator = await asyncIterator.next();
+
+            return {
+                index,
+                iterator
+            };
+        }
+        catch (err) {
+            if (options.errorCallback) {
+                options.errorCallback(err, index);
+            }
+            if (options.throwError !== false) {
+                return Promise.reject(err);
+            }
+
+            return {
+                index,
+                iterator: {
+                    done: true
+                }
+            };
+        }
+    };
 }
 
 async function* combineAsyncIterators(...iterators) {
+    let [options] = iterators;
+
+    if (typeof options.next === "function") {
+        options = {};
+    }
+    else {
+        iterators.shift();
+    }
     // Return if iterators is empty (avoid infinite loop).
     if (iterators.length === 0) {
         return;
     }
-    const promiseThatNeverResolve = new Promise(() => null);
+    const getNextAsyncIteratorValue = getNextAsyncIteratorFactory(options);
+
 
     try {
-        const asyncIteratorsValues = iterators.map(getNextAsyncIteratorValue);
+        const asyncIteratorsValues = new Map();
+        iterators.forEach((it, idx) => asyncIteratorsValues
+            .set(idx, getNextAsyncIteratorValue(it, idx))
+        );
         let numberOfIteratorsAlive = iterators.length;
 
         do {
-            const { iterator, index } = await Promise.race(asyncIteratorsValues);
-            if (iterator.done) {
-                numberOfIteratorsAlive--;
-                // We dont want Promise.race to resolve again on this index
-                // so we replace it with a Promise that will never resolve.
-                asyncIteratorsValues[index] = promiseThatNeverResolve;
-            }
-            else {
-                yield iterator.value;
-                asyncIteratorsValues[index] = getNextAsyncIteratorValue(iterators[index], index);
+            const response = await Promise.race(
+                Array.from(asyncIteratorsValues.values())
+            );
+            if (response) {
```

Comment:
Why checking for response here ? For me race in this case must always return a valid value because of the try/catch in getNextAsyncIteratorFactory.

So the code may be:
```js
const { iterator, index } = await Promise.race(
```